src/util/mpir_thread.c

```diff
@@ -0,0 +1,99 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+
+#include <mpiimpl.h>
+
+#ifdef MPICH_IS_THREADED
+/* use MPIR_Add_mutex to register mutex, they will be created/destroyed together
+ * at init/finalize */
+
+#define MAX_MUTEX_COUNT 64
+static MPID_Thread_mutex_t *mutex_list[MAX_MUTEX_COUNT];
+static int mutex_count = 0;
+
+void MPIR_Add_mutex(MPID_Thread_mutex_t * p_mutex)
+{
+    MPIR_Assert(mutex_count < MAX_MUTEX_COUNT);
+    mutex_list[mutex_count++] = p_mutex;
+}
+
+/* All the other mutexes that depend on the configured thread granularity */
+/* FIXME: we currently have  various mutexes scattered throughout the code
+ * and their logic interweaved between thread model and objects/features.
```

Comment:
Typos

Reply:
which word?

Reply:
Interweaved is a word sorry. I thought you meant interleaved